File: predictPartyVictory.py
from collections import deque
class Solution:
# Approach 1- Two Queues 
    def predictPartyVictory(self, senate: str): # TC O(n) // SC O(n)
        n=len(senate)
        rQueue=deque()
        dQueue=deque()
        for i, c in enumerate(senate):
            if c=="R":
                rQueue.append(i)
            else:
                dQueue.append(i)
        
        while rQueue and dQueue:
            rTurn=rQueue.popleft()
            dTurn=dQueue.popleft()
            
            if dTurn<rTurn:
                dQueue.append(dTurn+n)
            else:
                rQueue.append(rTurn+n)
        
        return "Radiant" if rQueue else "Dire"
            

s=Solution()
print(s.predictPartyVictory("RDDR"))

File: predictPartyVictory.java
import java.util.*;
// Approach 1- Two Queues
class Solution { // TC O(n) SC O(n)
    public String predictPartyVictory(String senate) {
        
        int n = senate.length();

        Queue<Integer> rQueue = new LinkedList<>();
        Queue<Integer> dQueue = new LinkedList<>();

        
        for (int i = 0; i < n; i++) {
            if (senate.charAt(i) == 'R') {
                rQueue.add(i);
            } else {
                dQueue.add(i);
            }
        }

        while (!rQueue.isEmpty() && !dQueue.isEmpty()) {
            
            int rTurn = rQueue.poll();
            int dTurn = dQueue.poll();

            if (dTurn < rTurn) {
                dQueue.add(dTurn + n);
            } else {
                rQueue.add(rTurn + n);
            }
        }

        return !rQueue.isEmpty() ? "Radiant" : "Dire";
    }
}

class predictPartyVictory{
    public static void main(String[] args) {
        String senate="RDD";
        Solution s= new Solution();
        System.out.println(s.predictPartyVictory(senate));
    }
}